prov/shm/src/smr_ep.c

```diff
@@ -1284,9 +1304,243 @@ static void smr_init_ipc_socket(struct smr_ep *ep)
+		owner_entry = container_of(*rx_entry, struct smr_rx_entry, peer_entry);
+		smr_entry = smr_get_recv_entry(srx_ctx->ep, owner_entry->iov, owner_entry->desc,
+					     owner_entry->peer_entry.count, addr,
+					     owner_entry->peer_entry.context,
+					     owner_entry->peer_entry.tag,
+					     owner_entry->ignore,
+					     owner_entry->peer_entry.flags & (~FI_MULTI_RECV));
+		if (!smr_entry) {
+			ret = -FI_ENOMEM;
+			goto out;
+		}
+
+		if (smr_adjust_multi_recv(srx_ctx, &owner_entry->peer_entry, size))
+			dlist_remove(dlist_entry);
+
+		smr_entry->peer_entry.owner_context = owner_entry;
+		*rx_entry = &smr_entry->peer_entry;
+		ofi_atomic_inc32(&owner_entry->ref);
+	} else {
+		dlist_remove(dlist_entry);
+	}
+
+	(*rx_entry)->srx = srx;
+	ret = FI_SUCCESS;
+out:
+	ofi_mutex_unlock(&srx_ctx->lock);
+	return ret;
+}
+
+static int smr_get_tag(struct fid_peer_srx *srx, fi_addr_t addr,
+			uint64_t tag, struct fi_peer_rx_entry **rx_entry)
+{
+	struct smr_rx_entry *smr_entry;
+	struct smr_srx_ctx *srx_ctx;
+	struct smr_match_attr match_attr;
+	struct dlist_entry *dlist_entry;
+	int ret;
+
+	srx_ctx = srx->fid.context;
+	ofi_mutex_lock(&srx_ctx->lock);
+
+	match_attr.id = addr;
+	match_attr.tag = tag;
+
+	dlist_entry = dlist_find_first_match(&srx_ctx->trecv_queue.list,
+					     srx_ctx->trecv_queue.match_func,
+					     &match_attr);
+	if (!dlist_entry) {
+		smr_entry = smr_alloc_rx_entry(srx_ctx);
+		if (!smr_entry) {
+			ret = -FI_ENOMEM;
+		} else {
+			smr_entry->peer_entry.owner_context = NULL;
+			smr_entry->peer_entry.addr = addr;
+			smr_entry->peer_entry.tag = tag;
+			smr_entry->peer_entry.srx = srx;
+			*rx_entry = &smr_entry->peer_entry;
+			ret = -FI_ENOENT;
+		}
+		goto out;
+	} else {
+		dlist_remove(dlist_entry);
+	}
+
+	*rx_entry = (struct fi_peer_rx_entry *) dlist_entry;
+	(*rx_entry)->srx = srx;
+	ret = FI_SUCCESS;
+out:
+	ofi_mutex_unlock(&srx_ctx->lock);
+	return ret;
+}
+
+static int smr_queue_msg(struct fi_peer_rx_entry *rx_entry)
+{
+	struct smr_srx_ctx *srx_ctx = rx_entry->srx->fid.context;
+
+	ofi_mutex_lock(&srx_ctx->lock);
+	dlist_insert_tail((struct dlist_entry *) rx_entry,
+			  &srx_ctx->unexp_msg_queue.list);
+	ofi_mutex_unlock(&srx_ctx->lock);
+	return 0;
+}
+
+static int smr_queue_tag(struct fi_peer_rx_entry *rx_entry)
+{
+	struct smr_srx_ctx *srx_ctx = rx_entry->srx->fid.context;
+
+	ofi_mutex_lock(&srx_ctx->lock);
+	dlist_insert_tail((struct dlist_entry *) rx_entry,
+			  &srx_ctx->unexp_tagged_queue.list);
+	ofi_mutex_unlock(&srx_ctx->lock);
+	return 0;
+}
+
+static void smr_free_entry(struct fi_peer_rx_entry *entry)
+{
+	struct smr_srx_ctx *srx = (struct smr_srx_ctx *) entry->srx->fid.context;
+	struct smr_rx_entry *smr_entry, *owner_entry;
+
+	ofi_mutex_lock(&srx->lock);
+	smr_entry = container_of(entry, struct smr_rx_entry, peer_entry);
+	if (entry->owner_context) {
+		owner_entry = container_of(entry->owner_context,
+					   struct smr_rx_entry, peer_entry);
+		ofi_atomic_dec32(&owner_entry->ref);
```

Comment:
Renaming ref to multi_recv_ref would clarify that the reference counting only applies to multi-recv buffers.